h/auth.py

```diff
@@ -0,0 +1,231 @@
+# -*- coding: utf-8 -*-
+"""
+OAuth integration.
+
+Supported grant types
+---------------------
+
+- A client credentials grant using the session and authenticating the client
+  with cross-site request forgery tokens.
+
+Supported token types
+---------------------
+
+- JSON Web Tokens [jwt]_ as bearer tokens.
+
+.. [jwt] https://tools.ietf.org/html/draft-ietf-oauth-json-web-token
+
+Limitations
+-----------
+
+- No support for 3rd party clients exists yet.
+- No support for scopes yet.
+
+"""
+import datetime
+
+import jwt
+from oauthlib.common import generate_client_id
+from oauthlib.oauth2 import RequestValidator as _RequestValidator
+from pyramid.exceptions import BadCSRFToken
+from pyramid import session
+
+from pyramid.authentication import RemoteUserAuthenticationPolicy
+from pyramid.authentication import SessionAuthenticationPolicy
+from pyramid.authorization import ACLAuthorizationPolicy
+from pyramid.util import action_method
+
+from .interfaces import IClientFactory
+
+LEEWAY = 240  # allowance for clock skew in verification
+
+
+class RequestValidator(_RequestValidator):
+
+    """Validates JSON Web Tokens."""
+
+    def authenticate_client(self, request):
+        if request.client_id is None:
+            try:
+                session.check_csrf_token(request, token='assertion')
+            except BadCSRFToken:
+                return False
+            client_id = request.registry.settings['h.client_id']
+            client = get_client(request, client_id)
+            request.client = client
+            request.user = request.session.get('userid')
+        elif request.client_secret is not None:
+            client_id = request.client_id
```

Comment:
Yeah. I'm not in love with it. But the fact is that oauthlib has its own
request object and this allows us to build OAuth components that are not
pyramid specific.

Essentially what pyramid_oauthlib does is provide request properties so
that the pyramid request object can round trip through oauthlib components
rather than having the overhead of creating a separate request object. It
also means that in places where the integration is tricky, like getting
clients and sessions, we have access to the registry from the validator
callbacks. Beyond that, it just provides a way to use the registry to
provide implementations of grant types, tokens, etc.

So, it's not meant to be magical except where we need pyramid stuff. It's
supposed to quack like an oauthlib request object, which realizes all
parameters as attributes.
On Feb 19, 2015 11:06 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25014820>:
>
> > +class RequestValidator(_RequestValidator):
> > +
> > +    """Validates JSON Web Tokens."""
> > +
> > +    def authenticate_client(self, request):
> > +        if request.client_id is None:
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
>
> Ah, right. For my taste that's a bit too much magic just to save you from
> typing .params.get(...), but hey.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25014820>.
>